### ISSUE TITLE
Master GUI Refactoring, untangle the security measures

### ISSUE THREAD

**Issue by MagnaCapax** [MagnaCapax/PMSS#15]
13-14 years ago when it was initially built, the JS was encrypted, it was made as dynamic as possible and fetching parts of itself from pulsedmedia.com

Since JS has been viewable in all browsers for a really long time, and it's _"not that special"_ anymore, we should refactor it a little bit to make updates and development more convenient.

We should setup the data it fetches for automatic remote update in the static pool of content, and make the code easily viewable.

Only refactor and make easy to use, update urls at first -- this is very solid and robust piece of code, which unfortunately is unmaintainable.

This is prerequisite for other issues.

**Comment by MagnaCapax** [comment]
## Adversarial Review — Update 2

**Issue:** #15 — Master GUI Refactoring, untangle the security measures
**Reviewed:** 2026-04-15
**Prior reviews:** 2026-03-27 (NEEDS-INVESTIGATION), 2026-04-04 (REWORK)
**Updated verdict:** PROCEED — with two tracked gaps remaining

---

### Status Since Prior Review (2026-04-04)

Both concrete gaps called out in the prior review landed within 24 hours of that review:

| Commit | Issue | Gap resolved |
|--------|-------|--------------|
| `11a371a6` | #422 (closed) | `welcome.php` jQuery 1.3.2 → 1.12.4, matching `index.php` |
| `ac1877ab` | #423 (closed) | `static.pulsedmedia.com` CSS references removed from `welcome.php` and `info.php`; `pmssTabs.js` and `screen.css` bundled locally; `filesystem.php` updated to sync all three new assets; regression test added (`SkeletonWebLocalAssetTest.php`) |

`static.pulsedmedia.com` is now fully cleared from all production code. The only remaining references are in test assertions checking that the string is absent — correct guard placement.

Two further commits since #423:
- `a2320f4c` — `stats.php` Docker policy toggle fix (#425, orthogonal to #15)
- `de3c3e53` — serialized file readers unified into shared runtime helpers; `welcome.php` and `stats.php` simplified

The decomposition approach documented in prior reviews continues to work. The child issues landed cleanly and left no regressions.

---

### Adversarial Findings

**1. Worst case if implemented as-described today:** A Build agent reading the issue title "untangle the security measures" with no prior context would repeat the #349 pattern — treating the `eval(unserialize())` block at `index.php:38–42` as the security problem to remove. As documented in the 2026-04-04 review, the operator explicitly closed #349: the eval chain is the remote UI update mechanism, not a vulnerability to eliminate. The issue title continues to attract this failure mode.

**2. Inversion (maximum damage):** Remove the `eval+unserialize` block without a replacement distribution mechanism, silently killing PM's ability to push tab layout changes to all deployed installations from one endpoint. The local fallback keeps working, so the damage is invisible until someone tries to push a GUI tab update and nothing propagates.

**3. Regret test (6 months):** The issue will keep attracting misaligned filings as long as "untangle the security measures" is in the title. That phrase now describes settled, intentional architecture. The title is the remaining debt.

**4. Removal inversion:** If this meta issue were closed today, the remaining concrete work (filemanager CDN dependency tracking, ongoing decomposition) would need new homes. The issue still functions as a discovery pointer even if most of the original scope is done.

**5. Requirements challenge:** "Only refactor and make easy to use, update urls at first" — the URL update goal is now complete. The remaining refactoring work has no exit criterion. "Easy to use" is not a testable acceptance condition. `filemanager.php` at 4,072 lines is the obvious remaining target but has no dedicated tracking.

**6. Deletion opportunity:** `etc/skel/www/jquery.tabs-ie.css` (541 bytes) exists in the skeleton directory but is not referenced in any `skel/www` file and is not included in the `filesystem.php` sync list. It is orphaned — predates IE conditional removal (#347). Candidate for deletion.

---

### New Gaps Not Tracked Anywhere

**[GAP — filemanager.php external CDN dependency cluster]**
`filemanager.php` still loads 14+ external assets on every file manager page load:
- jQuery 3.7.1 from `ajax.googleapis.com` (lines 3447, 3712) — two loads, one slim, one full
- Bootstrap 4.5.0 from `stackpath.bootstrapcdn.com` (lines 3407, 3448, 3479, 3713) — multiple loads
- DataTables 2.0.8 from `cdn.datatables.net` (line 3714)
- Font Awesome 4.7.0 from `cdnjs.cloudflare.com` (line 3480)
- Dropzone 5.5.1 from `cdnjs.cloudflare.com` (lines 1170, 1210)
- ekko-lightbox 5.3.0 from `cdnjs.cloudflare.com` (lines 3481, 3715)
- highlight.js 10.0.3 from `cdnjs.cloudflare.com` (lines 3483, 3717)
- Ace editor 1.4.1 from `cdnjs.cloudflare.com` (line 3898)

None of these have a tracking issue. The #342 fix (jQuery XSS) addressed a vulnerability in the jQuery call sites but left the external CDN dependency on jQuery 3.7.1 in place. `filemanager.php` remains the largest external-CDN exposure surface in `skel/www`.

Note: `filemanager.php` is third-party-origin code. Bundling these dependencies would be a significant undertaking and may conflict with upstream maintenance. This needs a scoping decision before any agent attempts it.

**[GAP — jQuery still external for index.php and welcome.php]**
`index.php` (local fallback path, line 66) and `welcome.php` (line 56) both load jQuery 1.12.4 from `ajax.googleapis.com`. The `static.pulsedmedia.com` dependency is gone, but the Google CDN dependency remains for these two core pages. Whether to bundle jQuery locally alongside `pmssTabs.js` and `screen.css` is not tracked.

**[GAP — Chart.js still external in info.php]**
`info.php` line 32 loads Chart.js 4.4.0 from `cdn.jsdelivr.net`. This was present in the prior review's investigation results but no tracking issue was filed. The file now uses a local `screen.css` but still depends on external Chart.js.

**[MINOR — permission inconsistency in skel/www static assets]**
`jquery.tabs.css` and `jquery.tabs-ie.css` have the executable bit set (`rwxr-x---`). The newly bundled `pmssTabs.js` and `screen.css` are `rw-rw-r--`. Web-served static files do not need the executable bit. The inconsistency is harmless but worth normalizing during any future skel/www cleanup pass.

---

### Caller Analysis

No change from the 2026-03-27 review. The caller map for `index.php`, `checkGui.php`, and `filesystem.php` is unchanged. `filesystem.php` now correctly syncs three additional files (`jquery.tabs.css`, `pmssTabs.js`, `screen.css`).

---

### Investigation Results

**Current file state:**

| File | LOC | Change since 2026-04-04 |
|------|-----|------------------------|
| `etc/skel/www/index.php` | 280 | Unchanged since #423 |
| `etc/skel/www/welcome.php` | 1,142 | +~242 lines (new features; reader unification) |
| `etc/skel/www/stats.php` | 1,115 | +~275 lines (new features; Docker fix) |
| `etc/skel/www/filemanager.php` | 4,072 | Unchanged |
| `etc/skel/www/info.php` | 103 | `static.pulsedmedia.com` ref removed |
| `etc/skel/www/pmssTabs.js` | 44 | **New** — bundled tabs implementation |
| `etc/skel/www/screen.css` | 154 | **New** — bundled layout stylesheet |

**Observation:** `welcome.php` and `stats.php` are growing as features are added alongside decomposition. This is expected during active development — characterization tests are being added to preserve behavior as structure is extracted. Growth is feature-driven, not entropy-driven.

**eval+unserialize:** `index.php:38–42` is unchanged. Operator-confirmed intentional architecture. Not an open question.

---

### Recommendations

1. **Rename or reframe this issue.** "Untangle the security measures" now describes architecture the operator has decided to keep. A title such as "GUI dependency modernization: bundle local assets, remove external CDN dependencies" matches the remaining work and prevents future eval-removal filings.

2. **File a dedicated tracking issue for `filemanager.php` CDN dependencies.** The 14+ external CDN loads are the largest remaining external dependency surface. A scoping decision (vendor file vs. first-party maintenance) is needed before any agent attempts bundling.

3. **File a child issue for jQuery bundling in `index.php` / `welcome.php`.** A small, well-scoped change that follows directly from the pattern established by #423.

4. **Consider closing #15** once the above tracking issues are filed. The URL modernization goal (original stated scope) is complete. Keeping the issue open invites continued misaligned filings due to the title.

5. **`jquery.tabs-ie.css` is a safe deletion candidate.** Not referenced in any current `skel/www` file, not synced to users by `filesystem.php`. Remove in the next cleanup pass.

---

### Worth-Implementing Analysis

- **URL update goal:** DONE. `static.pulsedmedia.com` fully cleared. Test guards in place.
- **Code readability goal:** IN PROGRESS. Commit-by-commit decomposition with characterization test coverage. The approach is working.
- **Remaining CDN gaps:** LOW cost for jQuery bundling in `index.php`/`welcome.php` (follows #423 pattern). HIGH cost for `filemanager.php` (vendor code, 14+ dependencies, scoping decision required first).
- **Overall verdict:** PROCEED — the issue has been and continues to be a productive tracking container. The remaining work is concrete and appropriately sized. No architectural blockers. The only action needed on the issue itself is a title update to prevent recurrence of the #349 failure pattern.

---

### Anti-Pattern Hunt

- **God file:** `filemanager.php` at 4,072 lines — PHP logic, HTML, CSS, and JS interleaved. Largest file in `skel/www` by far. Not tracked for decomposition.
- **Duplicate external loads:** `filemanager.php` loads jQuery 3.7.1 twice (line 3447 slim, line 3712 full) and Bootstrap 4.5.0 three times across conditional render paths. Worth auditing for dead load paths.
- **Version sprawl:** Two jQuery versions now remain across `skel/www` — 1.12.4 (index.php + welcome.php), 3.7.1 (filemanager.php). Acceptable given filemanager.php's separate vendor origin, but worth documenting.
- **Orphaned file:** `jquery.tabs-ie.css` in `skel/www` — not synced to users, not referenced. Dead weight.
- **Executable bit on CSS:** `jquery.tabs.css` and `jquery.tabs-ie.css` have the executable bit set. Harmless but inconsistent with new bundled files.

---

### Decomposition Targets

Ongoing decomposition is working. No change to the target list from the 2026-03-27 review. Current priority order based on size and external dependency exposure:

1. **[CDN]** `filemanager.php` external CDN loads — needs scoping issue first (vendor origin)
2. **[BUNDLE]** jQuery 1.12.4 bundling for `index.php` / `welcome.php` — small, follows #423 pattern exactly
3. **[BUNDLE]** Chart.js bundling for `info.php` — small, contained change
4. **[DEAD-CODE]** `jquery.tabs-ie.css` — orphaned, remove in next cleanup pass
5. **[REFACTOR]** `filemanager.php` structural decomposition — long-term, tracked separately

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Adversarial Review — Update 3

**Issue:** #15 — Master GUI Refactoring, untangle the security measures
**Reviewed:** 2026-04-22
**Prior reviews:** 2026-03-27 (NEEDS-INVESTIGATION), 2026-04-04 (REWORK), 2026-04-15 (PROCEED)
**Updated verdict:** BLOCKED — #466 must land first

---

### Status Since Prior Review (2026-04-15)

Two commits landed since the prior review, both unrelated to this issue:

- `693df1a9` — Deluge managed core.conf limits refresh
- `3292e954` — preserve live rtorrent on SCGI timeout (Refs #431)

No `skel/www` changes since 2026-04-14. None of the three tracking recommendations from the prior review have been acted on (issue title not updated, no filemanager.php CDN tracking issue filed, no jQuery bundling issue filed).

---

### Critical New Finding: #466 (filed 2026-04-18)

**Every `complete-verify` child issue of #15 is effectively NOT running on the fleet.**

Issue #466 documents that the remote `guiFrames.php` self-updater — the `eval(unserialize(base64_decode()))` pipeline at `index.php:38–42` — runs a file-revert loop on every page load after a PMSS update. The evaluatedcode fetches a `$fileVersions` table with SHA1s baked to OLD website versions, compares them to the locally deployed files, and overwrites any file that does not match the expected (old) SHA1.

Affected files that are actively being reverted on first post-update page load:

> `index.php`, `welcome.php`, `stats.php`, `deluge.php`, `info.php`, `lighttpdRestart.php`, `qbittorrent.php`, `rclone.php`, `rtorrentRestart.php`

This means:

| Issue | Marked | Operational status |
|-------|--------|-------------------|
| #347 — jQuery 1.4.1 cleanup | `complete-verify` | Reverted to jQuery 1.4.1 after first page load |
| #422 — welcome.php jQuery 1.3.2 sync | `complete-verify` | Reverted |
| #423 — bundle local assets | `complete-verify` | Reverted (local `pmssTabs.js`/`screen.css` are served but `index.php` itself loads the old JS) |

The `complete-verify` label reflects the repo state, not the fleet state. The revert is silent: no error log, no watchdog detection (checkGui.php only restores zero-byte/missing files, not SHA1-mismatched files).

---

### Adversarial Findings

**1. Worst case if #15 continues as-described:** A Build agent adds more `skel/www` improvements. Each lands on `git/main`. Each gets a `complete-verify` label. Each gets silently reverted on every fleet installation's next page load. The gap between "what main contains" and "what users run" grows with every commit, invisibly.

**2. Inversion (maximum damage):** Continue the current pattern — mark child issues `complete-verify` based on repo state, without verifying fleet state via `.guilog` or SHA1 comparison. The fleet runs 2010-era jQuery 1.4.1 indefinitely while the repo shows clean modern code.

**3. Regret test (6 months):** Six months of `skel/www` decomposition work produces zero user-facing improvements because every commit is reversed within 3 hours of a page load. The institutional record says the work was done; the user experience says it wasn't.

**4. Removal inversion:** If `PMSS_DISABLE_REMOTE_FRAMES` were already set to `1` in the lighttpd template and someone proposed removing it, we would fight to keep it — because removing it re-exposes the revert mechanism. This is the correct default.

**5. Requirements challenge:** The original requirement "Only refactor and make easy to use, update urls at first" is complete in the repo. The actual requirement that wasn't stated: changes must REACH THE FLEET. The missing delivery mechanism is the real work remaining.

**6. Deletion opportunity (the short-term fix):** `etc/seedbox/config/template.lighttpd` already has a `bin-environment` block at line 127. Adding `"PMSS_DISABLE_REMOTE_FRAMES" => "1"` to that block is a one-line change. `index.php:27` already reads this via `getenv()`. The local fallback path at `index.php:60–155` is complete: jQuery 1.12.4 from Google CDN, bundled `pmssTabs.js`, bundled `jquery.tabs.css`. The `setHeights()` viewport resize function is in `index.php`'s inline JS block (around line 245) and works correctly with jQuery 1.12.4. This is a genuine deletion opportunity — deleting the remote fetch path's operational effect with one config line.

---

### Why the Build Agent Keeps Failing on #466

Issue #466 has been blocked 8 times (attempts #2–8, all at 0 commits). The likely cause: the issue body presents three scopes in sequence — short-term (in-repo), medium-term (website coordination), long-term — and a sandboxed agent without network access attempts the full implementation, cannot reach `pulsedmedia.com/remote/guiFrames.php` to update SHA1s, and produces no output.

The short-term fix is entirely in-repo and can be implemented in a single commit:

```
etc/seedbox/config/template.lighttpd  — add PMSS_DISABLE_REMOTE_FRAMES to bin-environment
```

No network access required. No website coordination required. The env var is already supported in `index.php`. The local fallback is already complete and tested.

This scoping problem is the blocker, not the fix complexity. A dedicated child issue containing only the short-term scope is needed.

---

### Caller Analysis

No change from prior reviews. `checkGui.php` restores zero-byte/missing `index.php` only — does not detect SHA1-reverted files. `userConfigLighttpd.php:50` reads `template.lighttpd` and generates per-user configs during PMSS update runs. The template change propagates to users on the next update cycle.

---

### Investigation Results

**Current `index.php` state:**
- `eval+unserialize` block: `index.php:38–42` — unchanged, operator-architecture
- `PMSS_DISABLE_REMOTE_FRAMES` guard: `index.php:27` — already in place
- `setHeights()` function: inline JS block at `index.php:~245–260` — present, uses `$(window).height()`, runs on load and every 300ms
- Local fallback: `index.php:60–155` — uses jQuery 1.12.4 + `pmssTabs.js` + `jquery.tabs.css`

**`template.lighttpd` bin-environment block (lines 127–130):**
```
"bin-environment" => (
    "PHP_FCGI_CHILDREN" => "6",
    "PHP_FCGI_MAX_REQUESTS" => "5000"
),
```
One env var addition is the complete short-term fix.

**Test coverage:** `SkeletonWebLocalAssetTest.php` verifies local asset references and absence of retired CDN strings. No test currently asserts that `PMSS_DISABLE_REMOTE_FRAMES` is set in the lighttpd template.

---

### Recommendations

1. **BLOCK #15 on #466.** No further `skel/www` decomposition work reaches users until the revert mechanism is neutralised. Continuing decomposition without resolving #466 produces maintenance activity with zero user impact.

2. **File a scoped child issue for #466's short-term fix.** One commit: add `"PMSS_DISABLE_REMOTE_FRAMES" => "1"` to the `bin-environment` block in `etc/seedbox/config/template.lighttpd`. Include a test asserting the key is present. Codex can implement this without network access.

3. **After the lighttpd template fix lands:** re-verify the fleet state using the `.guilog` and SHA1 approach documented in #466. Confirm that post-update page loads no longer revert `index.php`.

4. **Then close or retitle #15.** Once the revert mechanism is off and fleet state is verified, resume the prior recommendations: retitle the issue, file the filemanager.php CDN tracking issue, file the jQuery bundling issue for `index.php`/`welcome.php`.

---

### Worth-Implementing Analysis

- **Cost of continued #15 work without #466 fix:** Negative ROI. Every commit disappears from the fleet on next page load.
- **Cost of #466 short-term fix:** One line in `template.lighttpd`, one test assertion. Approximately 15 minutes of Codex time.
- **Benefit of #466 short-term fix:** All prior and future `skel/www` work starts reaching users. jQuery 1.4.1 leaves the fleet. The revert surface (`eval+unserialize` operational effect) disappears.
- **Overall verdict:** BLOCKED. Unblock by landing the #466 short-term fix first. The #15 roadmap after that is clear.

---

### Anti-Pattern Hunt

- **FALSE_COMPLETION_CLAIM:** `complete-verify` labels on #347, #422, #423 reflect repo state, not fleet state. The fleet is running reverted versions. The claims are technically accurate (code is in the repo) but operationally misleading (code is not running). **FAIL** — flag for re-verification after #466 fix lands.
- **INVESTIGATION_LOOP_THEATER:** No stale re-investigation loops detected. Prior review findings hold.
- **REGEX_FOR_SEMANTICS:** Not applicable.
- **PREMATURE_OPTIMIZATION:** Not applicable.
- **OPERATOR_HANDOFF_ANTIPATTERN:** Not applicable.
- **FABRICATED_AUTHORITY_OR_PATH:** Not applicable.
- **FAIL_OPEN_SAFETY_BEHAVIOR:** The `getenv('PMSS_DISABLE_REMOTE_FRAMES')` check in `index.php:27` is correctly fail-open toward remote frames (the current default). Once the env var is set in the template, this becomes fail-closed toward local fallback. **Note:** until the template change propagates, individual servers remain in the old default. No silent failure — this is expected template propagation lag.
- **DRY_RUN_THEATER:** Not applicable.

---

### Decomposition Targets

No change from the 2026-04-15 review. All targets are unchanged and blocked on #466 resolution.

1. **[CDN]** `filemanager.php` external CDN loads — needs scoping issue first (vendor origin, ~14 external loads)
2. **[BUNDLE]** jQuery 1.12.4 in `index.php` / `welcome.php` local fallback — small, follows #423 pattern
3. **[BUNDLE]** Chart.js in `info.php` — small, contained
4. **[DEAD-CODE]** `jquery.tabs-ie.css` — orphaned, not synced by `filesystem.php`, not referenced
5. **[REFACTOR]** `filemanager.php` structural decomposition — long-term, requires dedicated tracking

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Adversarial Review — Update 4

**Issue:** #15 — Master GUI Refactoring, untangle the security measures
**Reviewed:** 2026-04-29
**Prior reviews:** 2026-03-27 (NEEDS-INVESTIGATION), 2026-04-04 (REWORK), 2026-04-15 (PROCEED), 2026-04-22 (BLOCKED)
**Updated verdict:** PROCEED — blocker resolved; three child issues needed for remaining work

---

### Status Since Prior Review (2026-04-22)

The blocker identified in Update 3 has been resolved.

| Item | Status |
|------|--------|
| #466 — guiFrames self-updater silently reverts skel/www files | **CLOSED** `complete-verify` (2026-04-27) |
| #471 — disable remote frames via lighttpd template env var | **CLOSED** `complete-verify` (2026-04-27) |
| `etc/seedbox/config/template.lighttpd:130` — `"PMSS_DISABLE_REMOTE_FRAMES" => "1"` | **Confirmed in source** (commit `b621a66b`) |
| `SkeletonWebLocalAssetTest.php:31-38` — test asserts env var present in template | **Confirmed in source** (commit `459af243`) |

The per-user lighttpd `bin-environment` block now sets `PMSS_DISABLE_REMOTE_FRAMES=1` for all PHP-CGI processes. The local fallback branch in `index.php:60–155` is now the permanent active path on every new deployment. The `eval(unserialize(base64_decode()))` pipeline at `index.php:27–58` is inoperative for all installations that have received the template update.

---

### New Finding: jQuery CDN Dependency Urgency Upgraded

**This is the substantive new finding in this update.** The `PMSS_DISABLE_REMOTE_FRAMES=1` fix has a second-order effect that prior reviews did not need to reason about: the local fallback path is now the only path.

Before the fix, `index.php:66` was reached only when the remote frame fetch failed (timeout, connectivity, PHP error). In practice, that meant it was a cold-path for installations where remote frames were operational. After the fix, it is the hot path on every single page load:

```html
<!-- index.php:66 — NOW ACTIVE ON EVERY PAGE LOAD -->
<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
```

```html
<!-- welcome.php:56 — same dependency, same status -->
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
```

The prior reviews (Update 2, Update 3) listed jQuery bundling as a low-priority trailing gap. It is no longer trailing — it is the active external CDN dependency for the two highest-traffic pages in the GUI.

**Interaction with #487:** Issue #487 (open, `needs-investigation`) documents that the per-instance update loop in update-step2.php does not propagate `welcome.php` changes to previously-deployed installations. jQuery bundling in `welcome.php` will not reach previously-deployed installations until #487 is resolved. The two issues should land in coordination: the `welcome.php` jQuery fix gains delivery reach from #487's fix.

---

### Adversarial Findings

**1. Worst case if #15 continues as-is:** A Build agent re-reads the title "untangle the security measures," concludes the `eval(unserialize())` block is the remaining work, and removes it. While the remote frames path is now effectively dead (env var disables it), removing the code is a CALLER-ENUMERATION scope: `index.php:27–58` is not a simple dead-code block — it is the architecture the operator explicitly kept open in 2026-04-04 (closed #349 with confirmation it is intentional remote-update mechanism). The code should be removed only after the operator confirms the remote update architecture is being retired, not by an agent inferring retirement from "the env var disables it."

**2. Inversion (maximum damage):** Bundle a malformed `jquery.min.js` locally and push it via the `filesystem.php` sync — all deployed GUI pages break silently at jQuery load time. The same deploy pipeline that makes #423 work becomes the attack surface if the bundled file is wrong. Mitigation: the file content must be the canonical unmodified jQuery 1.12.4 from npm (verified checksum before commit).

**3. Regret test (6 months):** The three untracked CDN gaps continue to accumulate. The filemanager.php situation worsens: a compat patcher (`pmssUserApplySkeletonFiles`) is already patching jQuery URLs on previously-deployed installations. If a `filemanager.php` CDN vendor outage occurs, there is no local fallback — unlike `index.php`, which has the env-var-controlled local path. Filing the scoping issue now prevents a reactive scramble when a CDN goes down.

**4. Removal inversion:** If `PMSS_DISABLE_REMOTE_FRAMES=1` were already set in all templates and someone proposed removing it, we would fight hard to keep it — the fleet's deployment fidelity depends on it. This is now settled.

**5. Requirements challenge:** The original requirement "Only refactor and make easy to use, update urls at first" is complete in the repo. The remaining requirements that were never explicitly stated: (a) all external CDN dependencies replaced by local bundles, and (b) changes must actually reach previously-deployed installations. Neither is in the issue body. They should be.

**6. Deletion opportunity:** `etc/skel/www/jquery.tabs-ie.css` (541 bytes). Zero references anywhere in `skel/www`, `scripts/`, or test files. Not included in the `filesystem.php` user sync list (confirmed: `SkeletonWebLocalAssetTest.php:40-51` asserts `jquery.tabs.css`, `pmssTabs.js`, `screen.css` — no `jquery.tabs-ie.css`). Executable bit set (`rwxr-x---`), inconsistent with the `rw-rw-r--` permissions on `pmssTabs.js` and `screen.css`. Deletion is safe, zero blast radius.

---

### Caller Analysis

No change from prior reviews. The caller map for `index.php` (via `checkGui.php` and `filesystem.php`) is unchanged. The `filesystem.php` sync list now includes `jquery.tabs.css`, `pmssTabs.js`, and `screen.css` — confirmed by `SkeletonWebLocalAssetTest.php:40-51`. A jQuery bundle file (e.g., `jquery.min.js`) would need to be added to this list in the same commit that adds the file to `etc/skel/www/`.

---

### Investigation Results

**Current skel/www state after #466/#471:**

| File | External CDN references | Status |
|------|------------------------|--------|
| `index.php` | `ajax.googleapis.com/jquery/1.12.4` (line 66) | Active — permanent now |
| `welcome.php` | `ajax.googleapis.com/jquery/1.12.4` (line 56) | Active — delivery blocked by #487 |
| `info.php` | `cdn.jsdelivr.net/chart.js@4.4.0` (line 32) | Untracked |
| `filemanager.php` | 15 CDN lines across 6 domains (lines 1170–3898) | Vendor-origin, compat-patched, separate decision |
| `jquery.tabs-ie.css` | n/a | Orphaned, zero callers, executable bit |

**Test coverage state:**
- `SkeletonWebLocalAssetTest.php` now has 4 tests covering: screen.css bundled, tabs assets bundled, `PMSS_DISABLE_REMOTE_FRAMES` in template, `filesystem.php` sync list
- **Gap:** No test asserts that `index.php` or `welcome.php` do NOT load jQuery from `ajax.googleapis.com`. Adding `pmssAssertRepoFileNotContainsString('etc/skel/www/index.php', 'ajax.googleapis.com')` would close this gap alongside a jQuery bundling commit.

**filemanager.php architecture note:** `userFilemanagerCompatPatchTest.php` asserts the PRESENCE of Google CDN jQuery 3.7.1 URLs in filemanager.php. Bundling jQuery for filemanager.php would require updating this compat patcher and its tests. The filemanager.php CDN situation is structurally different from the index.php/welcome.php case and warrants a dedicated scoping decision.

---

### Recommendations

1. **File a child issue for jQuery bundling in `index.php` / `welcome.php`.** Pattern is established by #423. Commit should: add `jquery.min.js` to `etc/skel/www/`, update `filesystem.php` to sync it, replace CDN references in `index.php:66` and `welcome.php:56`, add a `pmssAssertRepoFileNotContainsString` assertion to `SkeletonWebLocalAssetTest.php`. Note dependency on #487 for `welcome.php` delivery to previously-deployed installations.

2. **File a child issue for Chart.js bundling in `info.php`.** Single-file change, no compat patcher involved. Low blast radius.

3. **File a child issue for `filemanager.php` CDN audit / scoping.** 15 external CDN lines across 6 providers. Vendor-origin code with a dedicated compat patcher. Bundling requires a scoping decision (vendor shim vs. first-party maintenance). The issue should define the decision boundary, not prescribe the outcome.

4. **Delete `jquery.tabs-ie.css` in the next cleanup pass.** Zero callers, zero references, orphaned since IE conditional removal (#347). Include permission normalization for `jquery.tabs.css` (`rwxr-x---` → `rw-rw-r--`) in the same commit.

5. **Coordinate jQuery bundling with #487.** The `index.php` jQuery fix is independent and can land immediately. The `welcome.php` jQuery fix lands after or alongside #487's per-instance update-step2.php loop, or its benefit will be invisible to previously-deployed installations.

---

### Worth-Implementing Analysis

- **Cost (jQuery bundling):** LOW — one file to add (`jquery.min.js`, ~30KB), two file edits (`index.php:66`, `welcome.php:56`), one test assertion added, one `filesystem.php` sync entry. Exact same shape as #423.
- **Benefit (jQuery bundling):** HIGH — eliminates the last always-active external CDN dependency on the two highest-traffic pages. Each deployed installation becomes fully offline-capable for its GUI pages after a PMSS update.
- **Cost (Chart.js bundling):** LOW — single file to add, one line edit in `info.php`.
- **Cost (filemanager.php scoping):** LOW (the issue itself) — the actual work may be HIGH depending on decision. The issue is worth filing now to start the scoping conversation.
- **Cost (jquery.tabs-ie.css deletion):** TRIVIAL — delete one 541-byte file and fix one permission.
- **Overall verdict:** PROCEED — the issue continues to function as a productive tracking container. The remaining concrete work is well-scoped and directly continues the pattern established by the child issues already closed.

---

### Anti-Pattern Hunt

- **REGEX_FOR_SEMANTICS:** PASS — not applicable to this issue.
- **PREMATURE_OPTIMIZATION:** PASS — all proposed work has a demonstrated delivered value (CDN elimination pattern proven by #423).
- **OPERATOR_HANDOFF_ANTIPATTERN:** PASS — all remaining tasks are in-repo, Codex-implementable. No execution pushed to operator.
- **FALSE_COMPLETION_CLAIM:** PASS (post-#466 fix) — the `complete-verify` labels on prior child issues were accurate for repo state; the revert mechanism that made them misleading is now disabled. Re-verification of fleet state is warranted via the `.guilog` approach documented in #466.
- **FABRICATED_AUTHORITY_OR_PATH:** PASS — all file paths and line numbers cited above are verified from source.
- **FAIL_OPEN_SAFETY_BEHAVIOR:** PASS — `PMSS_DISABLE_REMOTE_FRAMES=1` in the template means no new deployments can revert to the remote frames path without explicit template modification. Installations that have not yet received the template update yet remain in the old default — this is expected propagation lag, not silent fail-open behavior.
- **DRY_RUN_THEATER:** PASS — not applicable.
- **INVESTIGATION_LOOP_THEATER:** PASS — no stale re-investigation loops. Each update has had new verified findings.

---

### Decomposition Targets

Priority order based on urgency and implementation cost:

1. **[BUNDLE-URGENT]** `etc/skel/www/index.php:66` and `etc/skel/www/welcome.php:56` — jQuery 1.12.4 from `ajax.googleapis.com`. Now always-active; no longer fallback-only. Bundle locally, follows #423 pattern exactly. Coordinate `welcome.php` delivery with #487.
2. **[BUNDLE]** `etc/skel/www/info.php:32` — Chart.js 4.4.0 from `cdn.jsdelivr.net`. Small, contained, no compat patcher involved.
3. **[CDN-SCOPE]** `etc/skel/www/filemanager.php` lines 1170–3898 — 15 external CDN lines across 6 providers. Requires scoping decision (vendor shim vs. first-party). File the scoping issue first.
4. **[DEAD-CODE]** `etc/skel/www/jquery.tabs-ie.css` — orphaned, not synced, no callers. Safe to delete.
5. **[PERMISSION]** `etc/skel/www/jquery.tabs.css` — executable bit (`rwxr-x---`) inconsistent with newer bundled files (`rw-rw-r--`). Normalize in same commit as #4.

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Investigation Report — Update 5

**Issue:** #15 — Master GUI Refactoring, untangle the security measures
**Reviewed:** 2026-05-31
**Prior reviews:** 2026-03-27 (NEEDS-INVESTIGATION), 2026-04-04 (REWORK), 2026-04-15 (PROCEED), 2026-04-22 (BLOCKED on #466), 2026-04-29 (PROCEED)
**Verdict:** SUBSTANTIALLY COMPLETE — recommend close-as-done (or retitle to a narrow tracker); do not send to a Build agent

### Status Since Prior Review (2026-04-29)
The central concern this issue describes has now been resolved at the root, by changes that postdate every prior review:

- **`ec38c41b` (Refs #601, 2026-05-28)** removes the remote `guiFrames` block from `etc/skel/www/index.php`. The diff deletes `$framesUrl = 'https://pulsedmedia.com/remote/guiFrames.php?v=2'`, `@base64_decode(...)`, `@unserialize($decoded)` and `eval($framesCode)`. **#601 is closed.**
- Current `index.php` (521 lines) sets `$useLocalFrames = true` and builds a static, local tab set. Verified: **0** occurrences of `eval`, `unserialize`, `base64_decode`, or `guiFrames`.
- The supply-chain "untangle the security measures" target — untrusted remote code deserialized and `eval`'d on every page load — no longer exists.

### What The Issue Asks For
From the body (filed Feb 2024): the master-GUI JS was historically encrypted and "made dynamic, fetching parts of itself from pulsedmedia.com." The ask: refactor so updates/development are convenient; move the remotely-fetched data into a static content pool; make the code easily viewable; **"only refactor and make easy to use, update urls at first"** — explicitly a narrow first step; stated as a **prerequisite for other issues**.

### Implementation Status
- **Remote-fetch / eval pipeline removed — DONE** (#601, verified by diff).
- **Code viewable/maintainable — DONE for `index.php`** (plain, commented PHP; named helpers; try/catch around the tabs plugin; `setHeights()` depends only on jQuery core). `welcome.php`/`stats.php` decomposition continues under **#316**.
- **Static content pool — DONE for the frame data; PARTIAL elsewhere.** Tab definitions are fully local. `welcome.php` still performs a *hardened, host-allow-listed* remote fetch (announcements RSS, restricted to `pulsedmedia.com`) — no `eval`; this is legitimately dynamic content rather than fully static.
- **URL modernization — DONE** (`ac1877ab`, Refs #423: removed `static.pulsedmedia.com`, bundled `pmssTabs.js`/`jquery.tabs.css` locally, added a regression test).
- **Security model documented — DONE via ADR-0016 (Customer-Facing PHP Tree Separation, inviolable) and ADR-0017 (Customer-Tree PHP Code-Review Checklist)**, both 2026-05-17. They codify the `etc/skel/www/` (customer) vs `/scripts/` (operator) boundary and a 6-point review checklist for any GUI change.
- **Remote-update reaching the fleet — DONE** (#466/#471, closed 2026-04-27: `PMSS_DISABLE_REMOTE_FRAMES=1` in the lighttpd template; #601 then removed the dead block).

### What's Missing (follow-on, not the "at first" scope)
- **External jQuery CDN on the two core pages** — `index.php:273` and `welcome.php:57` still load jQuery 1.12.4 from `ajax.googleapis.com`; no local bundle exists. Tracked by **#490** (open, verified still needed).
- **Orphaned `jquery.tabs-ie.css` + executable-bit normalization** — **#491** (open).
- **`filemanager.php`** — a ~4,072-line vendored Tiny File Manager drop-in with its own `shell_exec`/`exec` paths and ~14 external CDN loads. Third-party, structurally separate from the frame loader; **no dedicated tracking issue.**
- **`info.php`** — Chart.js from a CDN (minor, untracked).
- **Two ADR-0016 migrations still pending** — per ADR-0016's own migration record, two operator-tree helpers used by `welcome.php`/`stats.php` are not yet on the customer-read / operator-write-artifact pattern; until done, the GUI tree is not fully ADR-0016-compliant.

### Related Issues
- **#490** (open) — bundle jQuery locally for `index.php`/`welcome.php`; direct child, verified still needed; coordinate with **#487** for delivery to existing installs.
- **#491** (open) — orphan CSS + exec-bit cleanup; direct child, trivial.
- **#316** (open, active) — sustained decomposition of the customer frontends; the "make code maintainable" home beyond the frame loader.
- **#28** (open) — terminal window; **deferred by ADR-0015**, not part of #15.
- **#303 / #177** (open) — ruTorrent/rTorrent upgrades; embedded tab but a separate subsystem.
- **Closed children:** #347, #423, #466, #471, #601 (+ earlier #342/#357/#358/#374/#390/#422). The decomposition has been executed.

No two open issues modify the same lines in conflict; land #490 before/with #487.

### Related Regressions (customer-facing context)
The remote-frames mechanism was the root cause of a recurring user-panel regression class: blank/white page after login and intermittent 502s (the `frameData`-undefined family — #374/#466/#471), plus a separate broken tab-navigation regression (the `tabs-nav` class family — #536/#553) introduced by the local-asset bundling change, compounded by a self-updater that re-served old files until #471/#601. This recurred across multiple update cycles in early-to-mid 2026 with a low report-rate relative to affected users. With the remote-frames path removed, the silent-revert class is closed at its root; the cosmetic tab regressions were handled under #536/#553.

### ADR Adherence
- **ADR-0016** — `index.php` after #601 is clean customer-tree PHP, but ADR-0016's migration record lists two still-pending operator-tree helpers used by `welcome.php`/`stats.php`. Remaining GUI work must finish these via the prescribed operator-write-artifact / customer-read pattern (no punching the `/scripts/` boundary).
- **ADR-0017** — any remaining GUI change (#490/#491/filemanager) must pass the 6-point checklist.
- **ADR-0015** — governs the deferred browser console (#326/#28), not #15.

### Devil's Advocate / Deletion Analysis
- *Worst case if sent to a Build agent today:* it re-reads the title and tries to "remove the eval" that is already gone, or attempts a 6,000-line monolithic refactor. Every prior review reached the same "do not build this as a unit" conclusion.
- *Deletion (Musk Q2):* the dangerous block has already been deleted (#601) — best-part-is-no-part applied. The residual tracking is best **merged** into #316 (maintainability) and #490/#491 (CDN/cleanup), with one new issue for `filemanager.php`.
- *Null option:* if #15 is simply closed with nothing else, the only concrete loss is that `filemanager.php`'s vendored exec paths + CDN deps stay untracked — bounded to the file-manager tab, and mitigated by filing that one issue.

### Anti-Pattern Hunt
- **REGEX_FOR_SEMANTICS:** PASS — N/A.
- **PREMATURE_OPTIMIZATION:** PASS — remaining work follows the proven #423 pattern.
- **FALSE_COMPLETION_CLAIM:** PASS — the earlier "complete-verify reverted fleet-wide" concern is resolved at root (#471 disabled, #601 removed); each claim here is tied to a verified commit/diff or issue state.
- **OPERATOR_HANDOFF_ANTIPATTERN:** PARTIAL — the final close/retitle is left to the maintainer; justified because it is an administrative disposition on a long-lived tracking issue and one remnant needs an issue filed first.
- **FABRICATED_AUTHORITY_OR_PATH:** PASS — commit SHAs verified to exist; ADR contents read from source; the jQuery-external finding confirmed across multiple checks plus #490's open status.
- **FAIL_OPEN_SAFETY_BEHAVIOR:** PASS — `PMSS_DISABLE_REMOTE_FRAMES=1` means new deployments cannot revert to the remote path.
- **DRY_RUN_THEATER:** PASS — N/A.
- **INVESTIGATION_LOOP_THEATER:** PASS — this review adds genuinely new state (the #601 removal on 2026-05-28 and ADR-0016/0017 on 2026-05-17) and reaches a decisive disposition rather than another "PROCEED."

### Worth-Implementing Analysis
- The issue's literal scope (refactor the self-fetching frame loader into viewable static-local code, update URLs) is **done** (#601 + #423 + #347), and the security model is documented (ADR-0016/0017). Cost to "finish #15" = zero new code.
- The follow-on items are small, single-file, test-guarded changes (#490, #491, info.php) plus one scoping issue (`filemanager.php`). HIGH benefit for #490 (eliminates the last always-active external dependency on the two highest-traffic pages); LOW/contained cost.

### Recommendation
**Close #15 as substantially complete, or retitle it** (e.g. "GUI dependency modernization tracker") to stop the recurring misread where agents try to remove an `eval` that no longer exists. **Do not send it to a Build agent.** File one new issue to scope `filemanager.php` (vendored Tiny File Manager): decompose / bundle its ~14 external CDN deps and document its `shell_exec`/`exec` posture. Continue the remaining cleanups under the existing #490, #491, #316 and the ADR-0016 followups.

### Decomposition Targets (priority order)
1. **[BUNDLE]** jQuery 1.12.4 for `index.php:273` / `welcome.php:57` (#490) — last always-active external dep; follows the #423 pattern; coordinate `welcome.php` with #487.
2. **[BUNDLE]** Chart.js for `info.php` — small, contained.
3. **[SCOPE]** `filemanager.php` — vendored TFM; vendor-shim vs first-party decision needed (new issue).
4. **[DEAD-CODE]** `jquery.tabs-ie.css` + exec-bit normalization (#491).
5. **[MIGRATE]** the two pending ADR-0016 customer-tree helper migrations.

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Adversarial Review — Update 6 (correction of record)

**Issue:** #15 — Master GUI Refactoring, untangle the security measures
**Reviewed:** 2026-06-05
**Prior reviews:** 2026-03-27 (NEEDS-INVESTIGATION), 2026-04-04 (REWORK), 2026-04-15 (PROCEED), 2026-04-22 (BLOCKED), 2026-04-29 (PROCEED), 2026-05-31 (SUBSTANTIALLY COMPLETE)
**Verdict:** SHOULD-NOT-IMPLEMENT (meta-tracker, not Build-ready) — and the prior "the eval is gone" premise is now **false**

This is a focused correction triggered by genuinely new state since the last review, not a re-investigation. One material fact changed, and it inverts the foundation of the 2026-05-31 verdict.

### What changed (the correction)

The 2026-05-31 review concluded SUBSTANTIALLY COMPLETE on the premise that `ec38c41b` (#601, 2026-05-28) removed the remote `eval(unserialize(base64_decode()))` frame-loader and that `index.php` had "0 occurrences of eval / unserialize / base64_decode / guiFrames." **That premise no longer holds.**

- `13387808` (2026-06-03) **reverts #601** under an explicit operator directive: *"Local FAILOVER was supposed to be done, not remove the remote option completely."* The remote `guiFrames` fetch → `base64_decode` → `unserialize` → `eval` block is restored as the **PRIMARY** on-load path, with the local static frame set as **FAILOVER**.
- Current `etc/skel/www/index.php` (604 lines): guard `if (!getenv('PMSS_DISABLE_REMOTE_FRAMES'))` at line 305; remote fetch at 306–309; `eval($framesCode)` at line 315; every failure branch sets `$useLocalFrames = true` (failover to the local tab set).
- `13387808` also removed `PMSS_DISABLE_REMOTE_FRAMES=1` from `etc/seedbox/config/template.lighttpd` — remote frames are intentionally enabled again.
- Why #601's "dead code" premise was wrong: per the revert commit's operator-verified finding (production, 2026-06-03), the `PMSS_DISABLE_REMOTE_FRAMES` guard did not take effect on deployed installations, so the remote eval path was live fleet-wide the whole time. #601 therefore removed the **primary fleet-wide on-load GUI convergence path**, not dead code.

### Adversarial Findings

- **[DO-NOT-REMOVE] The remote/`eval` block is operator-confirmed intentional — twice.** #349 was closed (2026-03-22) stating the eval chain is the remote UI-update mechanism; #601's removal was then explicitly reverted (2026-06-03). The title "untangle the security measures" keeps baiting removal of this block. **Worst case (Munger inversion):** an implementer reads the now-stale "eval is gone / done" record plus the title, re-deletes the block, and silently kills central GUI distribution — the exact #601 defect, re-introduced. Removing it is operator-prohibited and breaks fleet convergence.
- **[FALSE_COMPLETION_CLAIM — corrected]** The prior "SUBSTANTIALLY COMPLETE / eval gone" is accurate only for the window 2026-05-28 → 2026-06-03. As of 2026-06-03 it is false. Flagged so the latest record is not read as current.
- **[meta-tracker, not Build-ready]** Unchanged across all six prior reviews: no acceptance criteria; conflates code-readability, URL modernization, and the remote-update security architecture. A Build agent cannot act on it as a unit.

### The actual remaining security work (operator decision, not a Build task)

The supply-chain surface is real and now confirmed-live: untrusted remote PHP is `eval`'d on each panel load; a compromise of the remote endpoint runs arbitrary code as each user's UID. But the operator wants the remote-update capability. The only mission-aligned direction is **harden, not remove**: replace `eval(unserialize(...))` with a signed / integrity-verified data manifest (no code execution), preserving remote updates without an RCE primitive. That is an operator-owned ADR decision, not in this issue's "at first" scope, and not a Build-agent task.

### Caller Analysis

No change from prior reviews. `index.php` reaches users via the GUI watchdog cron and the per-user update manifest (`scripts/lib/update/users/filesystem.php`); the per-user lighttpd config is rendered from `template.lighttpd` (`scripts/util/userConfigLighttpd.php:50`). The revert introduced no new callers.

### Anti-Pattern Hunt

- FALSE_COMPLETION_CLAIM: **FAIL (historical)** — prior "complete / eval-gone" claim is stale as of 2026-06-03; corrected above.
- INVESTIGATION_LOOP_THEATER: **PASS** — this comment adds exactly one new fact (the 2026-06-03 revert) and one structural action (the label), not a redundant re-analysis. The durable loop fix is #378, not another comment (see Recommendation).
- FAIL_OPEN_SAFETY_BEHAVIOR: noted — remote-primary / local-failover fails toward the local static set on any remote error (safe for availability); the RCE surface is the separate, operator-owned concern above.
- REGEX_FOR_SEMANTICS / PREMATURE_OPTIMIZATION / OPERATOR_HANDOFF_ANTIPATTERN / FABRICATED_AUTHORITY_OR_PATH / DRY_RUN_THEATER: N/A. All cited commit SHAs and issue numbers were verified against git / gh this session.

### Worth-Implementing Analysis

- **As a unit:** SHOULD-NOT-IMPLEMENT. Vague meta-tracker; its one concrete reading (remove the eval) is operator-prohibited.
- **Concrete follow-ons remain tracked:** #490 (bundle jQuery 1.12.4 locally — last always-active external CDN dep at `index.php:339` / `welcome.php:58`), #491 (orphan `jquery.tabs-ie.css` + exec-bit), #316 (sustained decomposition). `filemanager.php` (~4,072-line vendored file manager, ~14 external CDN loads, own `exec` paths) remains untracked → still warrants one dedicated scoping issue.

### Recommendation

1. **Applied `should-not-implement`.** This is the structural Build-agent skip and the precondition for #378 ("exclude should-not-implement from candidate pool"). The issue has cycled through seven automated passes; the missing label is part of why it keeps being re-selected.
2. **Operator disposition:** close as a tracker, or **retitle away from "untangle the security measures."** That phrase keeps baiting removal of an `eval` the operator has now twice chosen to keep.
3. Continue concrete work under #490 / #491 / #316; file one scoping issue for `filemanager.php` if none exists.

### Decomposition Targets

Unchanged from prior reviews: **[BUNDLE]** jQuery for `index.php` / `welcome.php` (#490); **[DEAD-CODE]** `jquery.tabs-ie.css` (#491); **[SCOPE]** `filemanager.php` vendored CDN cluster (untracked); **[REFACTOR]** sustained extraction (#316).

— Väinämöinen
